File: 백준/Gold/9663. N－Queen/N－Queen.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");
const N = Number(input[0]);
let result = 0;
const board = new Array(N).fill(0);
 
function nqueen(col, row, board) {
	for (let i = 0; i < col; i++) {
		if (board[i] === row) return false;
		if (Math.abs(col - i) === Math.abs(row - board[i])) return false;
	}
	return true;
}
 
function dfs(col, board) {
	const tmp = [...board];
	if (col === N) result++;
 
	for (let row = 0; row < N; row++) {
		if(nqueen(col, row, tmp)) {
			tmp[col] = row;
			dfs(col + 1, [...tmp]);
		}
	}
}
 
dfs(0, board);
console.log(result);

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
input = sys.stdin.readline

N = int(input())
result = 0
rows = [0] * N
diag1 = [0] * (2 * N - 1)
diag2 = [0] * (2 * N - 1)

def dfs(col):
    global result
    if col == N:
        result += 1
        return
    
    for row in range(N):
        if not rows[row] and not diag1[row - col + N - 1] and not diag2[row + col]:
            rows[row] = diag1[row - col + N - 1] = diag2[row + col] = 1
            dfs(col + 1)
            rows[row] = diag1[row - col + N - 1] = diag2[row + col] = 0

dfs(0)
print(result)